### ISSUE TITLE
No flycheck syntax checks in LSP mode after #2590

### ISSUE THREAD

**Issue by vlaci** [hlissner/doom-emacs#2636]
**What did you expect to happen?**
Running `flycheck-verify-setup` gives me the following result on 7cd9e55c8
![image](https://user-images.githubusercontent.com/1771332/75609691-19922f00-5b0b-11ea-93b6-8815cc24df2a.png)
Syntax errors are shown immediately after typing them using `lsp` checker
![image](https://user-images.githubusercontent.com/1771332/75611051-2452c100-5b17-11ea-8c79-f7b9081efc98.png)

**What actually happened?**
On c311d798c I see the following instead
![image](https://user-images.githubusercontent.com/1771332/75609833-0895ed80-5b0c-11ea-9eb6-71a6984cefe0.png)
Syntax errors are shown only after I have saved the buffer `python-pylint` and not shown for rust code where no additional checker is configured
![image](https://user-images.githubusercontent.com/1771332/75611077-6bd94d00-5b17-11ea-8fe8-c53a7fce4132.png)

**Steps to reproduce:**
1. Reproduction environment is created at https://github.com/vlaci/doom-emacs-issues/tree/master/2636
1. Minimal private configuration can be seen at https://github.com/vlaci/doom-emacs-issues/blob/master/2636/doom.d/init.el. The only change from `init.example.el` is that I have enabled `lsp` and `(python +lsp)` modules
1. `shell.nix` file is created for easy testing
    * `nix-shell -A good` to see the working case
    * `nix-shell -A bad` to see the erroneous case

**Note** the above example configurations are using unpinned emacs packages. I really hope that my issue does not depend on that quirk.

**System information:**
<details>

```
emacs   version    26.3
        features   XPM JPEG TIFF GIF PNG RSVG SOUND DBUS GSETTINGS GLIB NOTIFY LIBSELINUX GNUTLS LIBXML2 FREETYPE M17N_FLT LIBOTF XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS LIBSYSTEMD
        build      febr 29, 2020
        buildopts  (--prefix=/nix/store/gdcgg83innphpds09f19p2h0jkvk0fa7-emacs-26.3 --disable-build-details --with-modules --with-x-toolkit=gtk3 --with-xft CFLAGS=-DMAC_OS_X_VERSION_MAX_ALLOWED=101200)
        windowsys  x
        daemonp    server-running
doom    version    2.0.9
        build      fatal: not a git repository (or any of the parent directories): .git
        dir        /nix/store/b9k7nvmvv446r6f4ivhgwci3vz2qrk2s-doom-private/
system  type       gnu/linux
        config     x86_64-pc-linux-gnu
        shell      /nix/store/jirzjbbpl4j2kl2f01cfnrralxr5aw7a-bash-interactive-4.4-p23/bin/bash
        uname      Linux 5.5.5 #1-NixOS SMP Wed Feb 19 18:54:14 UTC 2020 x86_64
        path       (/nix/store/hm5y9ikxwaxlbfrbh1d8xh00nwr1vxjq-emacs-packages-deps/bin /nix/store/jirzjbbpl4j2kl2f01cfnrralxr5aw7a-bash-interactive-4.4-p23/bin /nix/store/ck9xy9qgpjhp3k80pwhlff9mq80w3ax5-patchelf-0.9/bin /nix/store/6lw4ansnm3dr08zf5nmcs9s92q9rhfxc-gcc-wrapper-9.2.0/bin /nix/store/va2z9a5k1ds65pxg7kfrs3p38zxk157k-gcc-9.2.0/bin /nix/store/ad115wnlpndz5xkkvz19k2ihsb5cdqpz-glibc-2.30-bin/bin /nix/store/3qgg8xzdyfd7nql0qlkb1xpnrfz4k1s2-coreutils-8.31/bin /nix/store/0j0h6cqglqd571lxci2xmhpq4z66s847-binutils-wrapper-2.31.1/bin /nix/store/7wy2mifb0igrp9hzf9dzwddadbs7ai7d-binutils-2.31.1/bin /nix/store/ad115wnlpndz5xkkvz19k2ihsb5cdqpz-glibc-2.30-bin/bin /nix/store/3qgg8xzdyfd7nql0qlkb1xpnrfz4k1s2-coreutils-8.31/bin /nix/store/91l4dlav09rqqg8bv2kbrd5nv3wf721r-doom-emacs/bin /nix/store/3qgg8xzdyfd7nql0qlkb1xpnrfz4k1s2-coreutils-8.31/bin /nix/store/cjxrwf292pxk5dp5fx737jn4bgmr10xc-findutils-4.7.0/bin /nix/store/j02iri0dpx6wp9fwzkgh7z31aqya80yr-diffutils-3.7/bin /nix/store/2kmapvyhr025v3cp3fdpnnz31q42knk7-gnused-4.8/bin /nix/store/9w5kl1scsqc7bw1hijm8s3j2s6d3lfas-gnugrep-3.4/bin /nix/store/8fanlnn1cb8ynn3p1d3lb1ylps94k1vi-gawk-5.0.1/bin /nix/store/x49zm66i7v2mcj71ppd459hmhllydapc-gnutar-1.32/bin /nix/store/4ib6spndkf8acwzp42xjygwbliq6j3h9-gzip-1.10/bin /nix/store/x9iaylq3f8r5y4z7bbgi76dnabh370bi-bzip2-1.0.6.0.1-bin/bin /nix/store/bpdpkkjpk2jmvjk6g08bjy1mvw14dyxd-gnumake-4.2.1/bin /nix/store/828g2nqfgivscv79xykkmgjk0znll08l-bash-4.4-p23/bin /nix/store/labk3cgvr3zygiamw2452s770j3r030p-patch-2.7.6/bin /nix/store/i7z20d9jnjjbz0ynlm2dfg672h4zjnvd-xz-5.2.4-bin/bin /nix/store/nafvbv9fw3fkgxp3fcbywf7smgjx11nz-kitty-0.16.0/bin /nix/store/r6ryjy02w3iac80rx2p30zs5hmmjk72n-imagemagick-6.9.10-71/bin /nix/store/lh8z803ya0lvbrv5gxc0nxm42s7i7gm5-xsel-unstable-2019-08-21/bin /nix/store/kiqr2dmcq5w2yhkis595n86y2y09a3ij-ncurses-6.1-20190112-dev/bin /run/wrappers/bin ~/.nix-profile/bin /etc/profiles/per-user/vlaci/bin /nix/var/nix/profiles/default/bin /run/current-system/sw/bin ~/.config/zsh/plugins/fast-syntax-highlighting /nix/store/gdcgg83innphpds09f19p2h0jkvk0fa7-emacs-26.3/libexec/emacs/26.3/x86_64-pc-linux-gnu)
config  envfile    nil
        elc-files  0
        modules    (:completion company ivy :ui doom doom-dashboard doom-quit hl-todo modeline nav-flash ophints (popup +all +defaults) vc-gutter vi-tilde-fringe window-select workspaces :editor (evil +everywhere) file-templates fold multiple-cursors rotate-text snippets :emacs dired electric ibuffer vc :checkers syntax :tools (eval +overlay) (lookup +docsets) lsp magit :lang data emacs-lisp markdown (org +dragndrop +present) (python +lsp) sh :config (default +bindings +smartparens))
        packages   (n/a)
        unpin      (n/a)
        elpa       (n/a)
```
</details>

**Comment by hlissner** [comment]
> Note the above example configurations are using unpinned emacs packages. I really hope that my issue does not depend on that quirk.

It quite possibly does. LSP has changed drastically in just the last week, and we hadn't caught up until #2590 was merged earlier today. These changes directly affected how flycheck (and company) were initiated in lsp-mode.

Try updating Doom and see if the issue is still present.

**Comment by hlissner** [comment]
> gives me the following result on 7cd9e55

Ah, but 7cd9e55 is just after the PR was merged, so maybe it isn't the issue? Can you reinstate pinning, however? Perhaps another upstream issue has broken flycheck integration.

**Comment by vlaci** [comment]
I have bisected to the offending commit I have mentioned above. The issue still persist after a traditional installation of doom-emacs (ac87afe)

<details>

```
emacs   version    26.3
        features   XPM JPEG TIFF GIF PNG RSVG SOUND DBUS GSETTINGS GLIB NOTIFY LIBSELINUX GNUTLS LIBXML2 FREETYPE M17N_FLT LIBOTF XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS LIBSYSTEMD
        build      febr 29, 2020
        buildopts  (--prefix=/nix/store/gdcgg83innphpds09f19p2h0jkvk0fa7-emacs-26.3 --disable-build-details --with-modules --with-x-toolkit=gtk3 --with-xft CFLAGS=-DMAC_OS_X_VERSION_MAX_ALLOWED=101200)
        windowsys  x
        daemonp    server-running
doom    version    2.0.9
        build      HEAD, origin/develop ac87afe34 2020-02-29 00:30:22 -0500
        dir        ~/.doom.d/
system  type       gnu/linux
        config     x86_64-pc-linux-gnu
        shell      /run/current-system/sw/bin/zsh
        uname      Linux 5.5.5 #1-NixOS SMP Wed Feb 19 18:54:14 UTC 2020 x86_64
        path       (/nix/store/nafvbv9fw3fkgxp3fcbywf7smgjx11nz-kitty-0.16.0/bin /nix/store/r6ryjy02w3iac80rx2p30zs5hmmjk72n-imagemagick-6.9.10-71/bin /nix/store/lh8z803ya0lvbrv5gxc0nxm42s7i7gm5-xsel-unstable-2019-08-21/bin /nix/store/kiqr2dmcq5w2yhkis595n86y2y09a3ij-ncurses-6.1-20190112-dev/bin /run/wrappers/bin ~/.nix-profile/bin /etc/profiles/per-user/vlaci/bin /nix/var/nix/profiles/default/bin /run/current-system/sw/bin ~/.config/zsh/plugins/fast-syntax-highlighting /nix/store/gdcgg83innphpds09f19p2h0jkvk0fa7-emacs-26.3/libexec/emacs/26.3/x86_64-pc-linux-gnu/)
config  envfile    envvar-file
        elc-files  0
        modules    (:completion company ivy :ui doom doom-dashboard doom-quit hl-todo modeline nav-flash ophints (popup +all +defaults) vc-gutter vi-tilde-fringe window-select workspaces :editor (evil +everywhere) file-templates fold multiple-cursors rotate-text snippets :emacs dired electric ibuffer vc :checkers syntax :tools (eval +overlay) (lookup +docsets) lsp magit :lang data emacs-lisp markdown (org +dragndrop +present) (python +lsp) sh :config (default +bindings +smartparens))
        packages   (n/a)
        unpin      (n/a)
        elpa       (n/a)
```
</details>

**Comment by vlaci** [comment]
I have updated the reproduction script to make sure it uses the referenced `lsp-mode` version. It still reproduces the issue as well as a 'proper' install of `doom-emacs`

**Comment by sebastiansturm** [comment]
before opening the lsp-backed buffer, could you call `(trace-function #'lsp-flycheck-enable)` to check whether `lsp-flycheck-enable` is called? Within the buffer, maybe evaluate `flycheck-checkers` and see if it contains `lsp`. @hlissner: I just noticed that with my changes to `tools/lsp/config.el`, `+lsp-init-flycheck-or-flymake-h` now calls `(lsp-flycheck-enable t)`. Which works but AFAICS should be just `(lsp-flycheck-enable)`; the argument is ignored

**Comment by sebastiansturm** [comment]
I installed nix and tried to use your reproduction script, but when opening test.py it doesn't even enable `lsp-mode`. Is `lsp` supposed to work out of the box (up to flycheck obviously), or do I need to do something else?

**Comment by vlaci** [comment]
You are right: `lsp-mode` does not get enabled. That seems to be the underlying issue. I was missing this, as messages buffer shows that the language server itself starts. Upon enabling `lsp-mode` by hand, syntax check works.

**Comment by sebastiansturm** [comment]
thanks for clearing that up. Though I don't know why the recent lsp PR should have broken auto-activation (python is even one of the languages I tried, and had no lsp-related issues so far). Does lsp-mode generate some helpful error messages? Or doesn't Emacs even try to activate it?

**Comment by vlaci** [comment]
It just won't get activated, tracing shown no attempts at calling `(lsp-mode)` during loading up a python file. I can enable it by hand without any errors also `(lsp)` itself gets called (hence the language server related related messages).

**Comment by vlaci** [comment]
i thought I known at least a little bit about elisp but I don't understand the following. Applying the following diff "fixes" the probllem:
```
diff --git a/modules/tools/lsp/config.el b/modules/tools/lsp/config.el
index 0b2e4f7da..292e7c6c5 100644
--- a/modules/tools/lsp/config.el
+++ b/modules/tools/lsp/config.el
@@ -139,7 +139,7 @@ This also logs the resolved project root, if found, so we know where we are."
                (and arg (not (equal arg 1))))))
          ;; read-process-output-max is only available on recent
          ;; development builds of Emacs 27 and above
-         (when (boundp 'read-process-output-max)
+         (when t
            (setq-local read-process-output-max (* 1024 1024)))
          (prog1 (lsp-mode 1)
            (setq-local lsp-buffer-uri (lsp--buffer-uri))
```

**Comment by sebastiansturm** [comment]
argh sorry, that is my fault. It's part of an `(and ...)`. I think the best thing would be to move the `(when (bound p) ...)` below `(lsp-mode 1)`. @hlissner: do you agree?

**Comment by vlaci** [comment]
Oh, I will never get used to the prefix logical operators :/ That's totally makes sense now. Thank you for helping me out. It's been a fun ride.

**Comment by sebastiansturm** [comment]
thanks for reporting this. Since I opened another PR I guess I should try it with an Emacs 26 installation this time :/

**Comment by hlissner** [comment]
Ah, I see the issue. I'll fix it shortly.

**Comment by hlissner** [comment]
And done! Should be fixed in 29650a6. GG0reat work figuring out the issue @vlaci ! I decided not to move it past the `(lsp-mode +1)`, because starting up the server could still benefit from the larger `read-process-output-max`, but it should work now, in any case. Let me know if that isn't the case and I'll reopen this.

@valrus

**Comment by vlaci** [comment]
I have verified that the issue is indeed fixed. Thank you!

**Comment by stratacast** [comment]
I can confirm that this error is still happening. On Ubuntu 19.10 with Emacs 26.3 and just built 27.0.90 to see if that helped since @vlaci mentioned 27 doesn't exhibit this behavior. The errors show up in the checker buffer yet not in the buffer that has my file. However, in Emacs 27 the errors remain in the flycheck buffer until I save

**Comment by SuperStevenZ** [comment]
I had a very similar issue which was fixed by enabling `:tools lsp` in `init.el`. 😃
